File: cpp/ctbpR.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <filesystem>
#include <chrono>
#include <iomanip>
#include <random>
#include <nlohmann/json.hpp> // You'll need to include this JSON library

using json = nlohmann::json;
namespace fs = std::filesystem;

class CursorResetError : public std::exception {
private:
    std::string message;
public:
    explicit CursorResetError(const std::string& msg) : message(msg) {}
    const char* what() const noexcept override { return message.c_str(); }
};

// Generate random hex string
std::string generateRandomHex(size_t length) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);
    std::stringstream ss;
    ss << std::hex << std::setfill('0');
    for (size_t i = 0; i < length; ++i) {
        ss << std::setw(2) << dis(gen);
    }
    return ss.str();
}

// Generate UUID (simplified version)
std::string generateUUID() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);
    std::stringstream ss;
    ss << std::hex;
    for (int i = 0; i < 8; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    ss << "-4";  // Version 4 UUID
    for (int i = 0; i < 3; i++) ss << dis(gen);
    ss << "-";
    ss << std::uniform_int_distribution<>(8, 11)(gen); // Variant
    for (int i = 0; i < 3; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 12; i++) ss << dis(gen);
    return ss.str();
}

fs::path getStorageFile() {
    std::string system = 
#ifdef _WIN32
        "Windows";
#elif __APPLE__
        "Darwin";
#elif __linux__
        "Linux";
#else
        throw CursorResetError("Unsupported operating system");
#endif

    fs::path base_path;
    if (system == "Windows") {
        base_path = fs::path(getenv("APPDATA")) / "Cursor";
    } else if (system == "Darwin") {
        base_path = fs::path(getenv("HOME")) / "Library" / "Application Support" / "Cursor";
    } else if (system == "Linux") {
        base_path = fs::path(getenv("HOME")) / ".config" / "Cursor";
    }
    return base_path / "User" / "globalStorage" / "storage.json";
}

fs::path backupFile(const fs::path& file_path) {
    if (!fs::exists(file_path)) return "";
    
    auto now = std::chrono::system_clock::now();
    auto in_time_t = std::chrono::system_clock::to_time_t(now);
    std::stringstream timestamp;
    timestamp << std::put_time(std::localtime(&in_time_t), "%Y%m%d_%H%M%S");
    
    fs::path backup_path = file_path.string() + ".backup_" + timestamp.str();
    fs::copy_file(file_path, backup_path);
    std::cout << "Created backup at: " << backup_path << std::endl;
    return backup_path;
}

json generateNewIds() {
    json new_ids;
    new_ids["machineId"] = generateRandomHex(32);
    new_ids["macMachineId"] = generateRandomHex(32);
    new_ids["devDeviceId"] = generateUUID();
    return new_ids;
}

void resetCursorId() {
    try {
        fs::path storage_file = getStorageFile();
        std::cout << "Targeting storage file: " << storage_file << std::endl;

        fs::create_directories(storage_file.parent_path());
        
        fs::path backup_path = backupFile(storage_file);
        
        json data;
        if (fs::exists(storage_file)) {
            std::ifstream in_file(storage_file);
            if (in_file.is_open()) {
                in_file >> data;
                in_file.close();
            }
        }

        json new_ids = generateNewIds();
        if (!data.contains("telemetry")) {
            data["telemetry"] = json::object();
        }
        data["telemetry"]["machineId"] = new_ids["machineId"];
        data["telemetry"]["macMachineId"] = new_ids["macMachineId"];
        data["telemetry"]["devDeviceId"] = new_ids["devDeviceId"];

        std::ofstream out_file(storage_file);
        if (!out_file.is_open()) {
            if (!backup_path.empty()) {
                fs::copy_file(backup_path, storage_file);
            }
            throw CursorResetError("Failed to write storage file");
        }
        out_file << std::setw(2) << data << std::endl;
        out_file.close();

        std::cout << "🎉 Device IDs have been successfully reset!" << std::endl;
        std::cout << "New device IDs:" << std::endl;
        std::cout << std::setw(2) << new_ids << std::endl;
    }
    catch (const CursorResetError& e) {
        std::cerr << "❌ Error: " << e.what() << std::endl;
        throw;
    }
    catch (const std::exception& e) {
        std::cerr << "❌ Unexpected error occurred: " << e.what() << std::endl;
        throw;
    }
}

int main() {
    std::cout << "Cursor Trial Reset Tool v2.0" << std::endl;
    std::cout << "---------------------------" << std::endl;
    
    try {
        resetCursorId();
        return 0;
    }
    catch (...) {
        std::cerr << "\nFor more details, check program output" << std::endl;
        return 1;
    }
}

File: py/ctbpR.py
#!/usr/bin/env python3
'''
Cursor Trial Reset Tool v2.0

This script resets the device IDs in Cursor's configuration file to generate new random IDs.
Optimized for Cursor v0.47.8

Repository: https://github.com/ultrasev/cursor-reset
Author: @ultrasev
Created: 10/Dec/2024
Updated: 23/Mar/2025
'''

import json
import os
import shutil
import uuid
import logging
from datetime import datetime
from pathlib import Path
import platform
from typing import Dict, Optional

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('cursor_reset.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

class CursorResetError(Exception):
    """Custom exception for Cursor reset operations"""
    pass

def backup_file(file_path: Path) -> Optional[Path]:
    """Create a timestamped backup of the given file."""
    try:
        if file_path.exists():
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            backup_path = file_path.with_suffix(f'.backup_{timestamp}')
            shutil.copy2(file_path, backup_path)
            logger.info(f"Created backup at: {backup_path}")
            return backup_path
        return None
    except Exception as e:
        logger.error(f"Failed to create backup: {str(e)}")
        raise CursorResetError(f"Backup creation failed: {str(e)}")

def get_storage_file() -> Path:
    """Determine the storage file location based on the operating system."""
    system = platform.system()
    base_path = {
        "Windows": Path(os.getenv("APPDATA", "")) / "Cursor",
        "Darwin": Path(os.path.expanduser("~")) / "Library" / "Application Support" / "Cursor",
        "Linux": Path(os.path.expanduser("~")) / ".config" / "Cursor"
    }.get(system)

    if not base_path:
        raise CursorResetError(f"Unsupported operating system: {system}")

    return base_path / "User" / "globalStorage" / "storage.json"

def generate_new_ids() -> Dict[str, str]:
    """Generate new device IDs."""
    return {
        "machineId": os.urandom(32).hex(),
        "macMachineId": os.urandom(32).hex(),
        "devDeviceId": str(uuid.uuid4())
    }

def reset_cursor_id() -> None:
    """Reset Cursor device IDs."""
    try:
        storage_file = get_storage_file()
        logger.info(f"Targeting storage file: {storage_file}")

        # Ensure directory exists
        storage_file.parent.mkdir(parents=True, exist_ok=True)
        
        # Create backup
        backup_path = backup_file(storage_file)

        # Load existing data or initialize new
        data = {}
        if storage_file.exists():
            try:
                with open(storage_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
            except json.JSONDecodeError:
                logger.warning("Invalid JSON in storage file, starting fresh")
                data = {}

        # Generate and apply new IDs
        new_ids = generate_new_ids()
        data["telemetry"] = data.get("telemetry", {})
        data["telemetry"].update({
            "machineId": new_ids["machineId"],
            "macMachineId": new_ids["macMachineId"],
            "devDeviceId": new_ids["devDeviceId"]
        })

        # Write updated data
        try:
            with open(storage_file, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            logger.info("Successfully updated storage file")
        except Exception as e:
            logger.error(f"Failed to write storage file: {str(e)}")
            # Restore backup if it exists
            if backup_path and backup_path.exists():
                shutil.copy2(backup_path, storage_file)
                logger.info("Restored backup due to write failure")
            raise

        # Print success message
        print("🎉 Device IDs have been successfully reset!")
        print("New device IDs:")
        print(json.dumps(new_ids, indent=2))

    except CursorResetError as e:
        logger.error(f"Reset failed: {str(e)}")
        print(f"❌ Error: {str(e)}")
        raise
    except Exception as e:
        logger.error(f"Unexpected error: {str(e)}")
        print(f"❌ Unexpected error occurred: {str(e)}")
        raise

def main():
    """Main entry point with basic argument handling."""
    try:
        print("Cursor Trial Reset Tool v2.0")
        print("---------------------------")
        reset_cursor_id()
    except Exception as e:
        logger.critical(f"Program terminated: {str(e)}")
        print("\nFor more details, check cursor_reset.log")
        exit(1)

if __name__ == "__main__":
    main()